### DISCUSSION TITLE
can you get storage row by id?

### DISCUSSION THREAD

**Discussion by lochie** [supabase/supabase#13604] [Questions]
i might be missing something, but is there a way to get a stored item from a bucket by an id?
only documentation i can find on getting an item is [here](https://supabase.com/docs/reference/javascript/storage-from-download), but i was expecting to find something like the below in the documentation since it seems pretty weird that the only way to get an file is by it's pathname 🤔

```tsx
const { data, error } = await supabase
  .storage
  .from(...)
  .download(uuid) // would be nice to do this
```

**Comment by abdirahmn1** [comment]
To anyone that might stumble on this in the future, this problem is already solved, here's the [PR](https://github.com/supabase/storage-api/pull/332) that was merged, a [changelog](https://supabase.com/changelog) was also published on july 7 2023.

**Reply by joeldomke** [reply]
I don't think this answers the question. The goal would be to download a file by specifying the object id instead of the pathname.

**Reply by GaryAustin1** [reply]
That is still not a feature.

**Reply by cohesivejones** [reply]
This would be really handy when you want to upload two files with the same name

**Comment by CorneliusSchramm** [comment]
I solved this by creating a view that joins my reference table with `storage.objects`, making it easier to query file paths from the public schema:

```sql
create view public.my_view as
select
  ref_table.entity_id,
  ref_table.storage_ref_id,
  s.bucket_id,
  s.name
from
  my_reference_table ref_table
  join storage.objects s on ref_table.storage_ref_id = s.id;
```

The reference table has a foreign key linking to `storage.objects`:

```sql
create table public.my_reference_table (
  id uuid not null default gen_random_uuid (),
  constraint my_reference_table_pkey primary key (id),
...
  storage_ref_id uuid null,
  constraint my_reference_table_storage_ref_id_fkey foreign key (storage_ref_id) references storage.objects (id)
);
```

Then, using the Supabase JS SDK, I fetch the path like this:

```javascript
const { data, error } = await supabase
  .from('my_view')
  .select('bucket_id, name')
  .eq('entity_id', your_entity_id);
```

This avoids querying `storage.objects` directly while keeping the structure clean.

**Comment by barrownicholas** [comment]
This is what we did:

```sql
create or replace function public."storage.objects.get_object_by_id"(object_id uuid)
returns storage.objects
language sql
security invoker
as $$
  select *
  from storage.objects
  where id = object_id
  limit 1;
$$;
```
